src/ccl_cls.c

```diff
@@ -0,0 +1,1336 @@
+    }
+  }
+  else {
+    int i,nchi=(int)((clt->chimax-clt->chimin)/w->dchi)+1;
+    for(i=0;i<nchi;i++) {
+      double chi=clt->chimin+w->dchi*(i+0.5);
+      if(chi<=clt->chimax) {
+	double a=ccl_scale_factor_of_chi(cosmo,chi,status);
+	double pk=ccl_nonlin_matter_power(cosmo,k,a,status);
+	double jl=ccl_j_bessel(l,k*chi);
+	double f_all=f_dens(a,cosmo,clt,status)*jl;
+	if(clt->has_rsd) {
+	  double ddjl,x=k*chi;
+	  if(x<1E-10) {
+	    if(l==0) ddjl=0.3333-0.1*x*x;
+	    else if(l==2) ddjl=-0.13333333333+0.05714285714285714*x*x;
+	    else ddjl=0;
+          }
+	  else {
+	    double jlp1=ccl_j_bessel(l+1,x);
+	    ddjl=((x*x-l*(l-1))*jl-2*x*jlp1)/(x*x);
+	  }
+	  f_all+=f_rsd(a,cosmo,clt,status)*ddjl;
+	}
+	if(clt->has_magnification)
+	  f_all+=-2*clt->prefac_lensing*l*(l+1)*f_mag(a,chi,cosmo,clt,status)*jl/(k*k);
+
+	ret+=f_all*sqrt(pk); //TODO: is it worth splining this sqrt?
+      }
+    }
+    ret*=w->dchi;
+  }
+
+  return ret;
+}
+
+static double f_lensing(double a,double chi,ccl_cosmology *cosmo,CCL_ClTracer *clt, int * status)
+{
+  double wL=ccl_spline_eval(chi,clt->spl_wL);
+
+  if(wL<=0)
+    return 0;
+  else
+    return clt->prefac_lensing*wL/(a*chi);
+}
+
+static double f_IA_NLA(double a,double chi,ccl_cosmology *cosmo,CCL_ClTracer *clt, int * status)
+{
+  if(chi<=1E-10)
+    return 0;
+  else {
+    double a=ccl_scale_factor_of_chi(cosmo,chi, status);
+    double z=1./a-1;
+    double pz=ccl_spline_eval(z,clt->spl_nz);
+    double ba=ccl_spline_eval(z,clt->spl_ba);
+    double rf=ccl_spline_eval(z,clt->spl_rf);
+    double h=cosmo->params.h*ccl_h_over_h0(cosmo,a,status)/CLIGHT_HMPC;
+
+    return pz*ba*rf*h/(chi*chi);
+  }
+}
+
+
+//Transfer function for shear
+//l -> angular multipole
+//k -> wavenumber modulus
+//cosmo -> ccl_cosmology object
+//w -> CCL_ClWorskpace object
+//clt -> CCL_ClTracer object (must be of the CL_TRACER_WL type)
+static double transfer_wl(int l,double k,
+			  ccl_cosmology *cosmo,CCL_ClWorkspace *w,CCL_ClTracer *clt, int * status)
+{
+  double ret=0;
+  if(l>w->l_limber) {
+    double chi=(l+0.5)/k;
+    if(chi<=clt->chimax) {
+      double a=ccl_scale_factor_of_chi(cosmo,chi,status);
+      double pk=ccl_nonlin_matter_power(cosmo,k,a,status);
+      double jl=j_bessel_limber(l,k);
+      double f_all=f_lensing(a,chi,cosmo,clt,status)*jl;
+      if(clt->has_intrinsic_alignment)
+	f_all+=f_IA_NLA(a,chi,cosmo,clt,status)*jl;
+
+      ret=f_all*sqrt(pk);
+    }
+  }
+  else {
+    int i,nchi=(int)((clt->chimax-clt->chimin)/w->dchi)+1;
+    for(i=0;i<nchi;i++) {
+      double chi=clt->chimin+w->dchi*(i+0.5);
+      if(chi<=clt->chimax) {
+	double a=ccl_scale_factor_of_chi(cosmo,chi,status);
+	double pk=ccl_nonlin_matter_power(cosmo,k,a,status);
+	double jl=ccl_j_bessel(l,k*chi);
+	double f_all=f_lensing(a,chi,cosmo,clt,status)*jl;
+	if(clt->has_intrinsic_alignment)
+	  f_all+=f_IA_NLA(a,chi,cosmo,clt,status)*jl;
+
+	ret+=f_all*sqrt(pk); //TODO: is it worth splining this sqrt?
+      }
+    }
```

Comment:
@rmjarvis: that is correct. This is a highly oscillatory function, and I have not found a GSL routine that solves it fast enough. It may be possible with https://www.gnu.org/software/gsl/doc/html/integration.html#qawo-adaptive-integration-for-oscillatory-functions tinkering with the integrand in a non-trivial way so one uses a standard integrator for low values of k and a sinusoidal approximation to the Bessel function at high k. However, at the time when we implemented this, we thought this wasn't worth the effort: the preferred non-Limber integrator in CCL is Angpow, and the main goal of this alternative Riemann integral is to use them in the Angpow unit tests (and to provide some support for certain things that Angpow doesn't do yet, such as magnification or shear power spectra).

In short, this power spectrum method is not advertised in the paper and is discouraged in the documentation, but we believe it's useful to keep it in the code for testing purposes. If you think this choice over-complicates things (and I may be inclined to agree), we could just create a benchmark file for the angpow unit tests and remove this code. Let me know how this sounds.

Reply:
Actually, ignore the above. In #506 we've got rid of the native non-Limber method for simplicity, so this is no longer an issue